cmd/ingest/ingest.go

```diff
@@ -63,22 +98,194 @@ type whitelistBaseEntity struct {
-	promulgatedId *charm.URL
-	// channel holds a channel where the entity has been published.
-	channel params.Channel
-	// current holds whether this revision is current for the above channel.
-	current bool
+func (e *whitelistBaseEntity) isBundle() bool {
+	for _, e := range e.entities {
+		return e.id.Series == "bundle"
+	}
+	return false
+}
+
+type IngestStats struct {
+	BaseEntityCount     int
+	EntityCount         int
+	FailedEntityCount   int
+	ArchivesCopiedCount int
+	Errors              []string
+}
+
+func ingest(p ingestParams) IngestStats {
+	ing := &ingester{
+		dest:    p.dest,
+		src:     p.src,
+		limiter: newLimiter(20),
+	}
+	resolvedEntities := ing.resolveWhitelist(p.whitelist)
+
+	// First transfer all charms.
+	for _, baseInfo := range resolvedEntities {
+		for _, entity := range baseInfo.entities {
+			entity := entity
+			if entity.id.Series != "bundle" {
+				// TODO perhaps we should consider limiting the number of goroutines we start here?
+				ing.limiter.do(func() {
+					ing.transferEntity(entity)
+				})
+			}
+		}
+	}
+	ing.limiter.wait()
+
+	// TODO Then transfer all resources that are either part of the published
+	// resources associated with a base entity, or are mentioned specifically
+	// by a bundle.
+	// we need to choose an entity to upload the resource for.
+	// get all resources for the base entity
+	// include resource if:
+	// 	it's currently published in one of the channels that we've specified
+	// 	it's mentioned in a bundle
+
+	// Then transfer all bundles (we have to do this after transferring
+	// charms, as we can't upload a bundle without uploading its
+	// charms first).
+	for _, baseInfo := range resolvedEntities {
+		for _, entity := range baseInfo.entities {
+			entity := entity
+			if entity.id.Series == "bundle" {
+				ing.limiter.do(func() {
+					ing.transferEntity(entity)
+				})
+			}
+		}
+	}
+	ing.limiter.wait()
+
+	// TODO transfer common-info for all entities.
+
+	return ing.stats(resolvedEntities)
+}
+
+// stats returns statistics about transferred charmstore entities.
+func (ing *ingester) stats(es map[string]*whitelistBaseEntity) IngestStats {
+	stats := IngestStats{
+		BaseEntityCount: len(es),
+		Errors:          ing.errors,
+	}
+	for _, baseEntity := range es {
+		stats.EntityCount += len(baseEntity.entities)
+		for _, e := range baseEntity.entities {
+			if !e.synced {
+				stats.FailedEntityCount++
+			}
+			if e.archiveCopied {
+				stats.ArchivesCopiedCount++
+			}
+		}
+	}
+	return stats
+}
+
+func (ing *ingester) transferEntity(e *entityInfo) {
+	// TODO retry a couple of times if this fails with a temporary-looking error?
+
+	// First find out whether the entity already exists in the destination charmstore.
+	// If so, we only need to transfer metadata.
+
+	// Use NoChannel which should pick a published channel for us, as we
+	// don't know what the destination has previously been published as. Sigh.
```

Comment:
as we have no destinations yet surely we could change charmstore to make this available?